File: java/leetcode_grind/Day585.java
package leetcode_grind;

import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.Map;

public class Day585 {
    // https://leetcode.com/problems/count-number-of-nice-subarrays/description
    static class Solution1 {
        public int numberOfSubarrays(int[] nums, int k) {
            int currSum = 0, subarrays = 0;
            Map<Integer, Integer> prefixSum = new HashMap<>();
            prefixSum.put(currSum, 1);

            for (int i = 0; i < nums.length; i++) {
                currSum += nums[i] % 2;
                if (prefixSum.containsKey(currSum - k)) {
                    subarrays += prefixSum.get(currSum - k);
                }
                prefixSum.put(currSum, prefixSum.getOrDefault(currSum, 0) + 1);
            }
            return subarrays;
        }
    }

    // https://leetcode.com/problems/count-number-of-nice-subarrays/description/
    static class Solution2 {
        public int numberOfSubarrays(int[] nums, int k) {
            int subarrays = 0, lastPopped = -1, initialGap = -1;
            var oddIndices = new ArrayDeque<Integer>();
            for (int i = 0; i < nums.length; i++) {
                if (nums[i] % 2 == 1) {
                    oddIndices.addLast(i);
                }

                if (oddIndices.size() > k) {
                    lastPopped = oddIndices.removeFirst();
                }

                if (oddIndices.size() == k) {
                    initialGap = oddIndices.peekFirst() - lastPopped;
                    subarrays += initialGap;
                }

            }
            return subarrays;
        }
    }

    // https://leetcode.com/problems/count-number-of-nice-subarrays/description
    static class Solution3 {
        public int numberOfSubarrays(int[] nums, int k) {
            int subrrays = 0, initialGap = 0, qsize = 0, start = 0;
            for (int end = 0; end < nums.length; end++) {
                if (nums[end] % 2 == 1) {
                    qsize++;
                }
                if (qsize == k) {
                    initialGap = 0;

                    while (qsize == k) {
                        qsize -= nums[start] % 2;
                        initialGap++;
                        start++;
                    }
                }
                subrrays += initialGap;
            }
            return subrrays;
        }
    }
}

File: src/leetcode_grind/day_585.rs
// https://leetcode.com/problems/count-substrings-without-repeating-character/description
pub fn number_of_special_substrings(s: String) -> i32 {
    let mut ans = 0;
    let mut freq = vec![0; 26];
    let s = s.chars().collect::<Vec<_>>();
    let mut start = 0;
    for end in 0..s.len() {
        freq[s[end] as usize - 'a' as usize] += 1;
        while freq[s[end] as usize - 'a' as usize] > 1 {
            freq[s[start] as usize - 'a' as usize] -= 1;
            start += 1;
        }
        ans += end - start + 1;
    }
    ans as i32
}

// https://leetcode.com/problems/count-number-of-nice-subarrays/description
pub fn number_of_subarrays(nums: Vec<i32>, k: i32) -> i32 {
    fn at_most(nums: &Vec<i32>, k: i32) -> i32 {
        let mut window_size = 0;
        let mut subarrays = 0;
        let mut start = 0;

        for end in 0..nums.len() {
            window_size += nums[end] % 2;
            while window_size > k {
                window_size -= nums[start] % 2;
                start += 1;
            }
            subarrays += end - start + 1;
        }
        subarrays as i32
    }

    at_most(&nums, k) - at_most(&nums, k - 1)
}
